### PR TITLE
metrics: Improve monitoring BPF maps and userspace caches

### LINKED ISSUE #1775: bpf maps:metrics: fix bpf maps metrics
### What happened?

Issues:

1. ~~Reading wrong bpf maps: we add the stats to the maps where it should already have it? this could lead to x_stats_stats~~ https://github.com/cilium/tetragon/blob/3326a78fbb120a9f5946b64733f8786d08be4b0e/pkg/observer/observer_stats.go#L33

2. ~~We fail or return here~~ https://github.com/cilium/tetragon/blob/3326a78fbb120a9f5946b64733f8786d08be4b0e/pkg/observer/observer_stats.go#L37  ~~and I guess we iterate over all maps~~

3. Rename all bpf map stats with "tg_$name_map_stats" schema

### Tetragon Version

master

### Kernel Version

None

### Kubernetes Version

_No response_

### Bugtool

_No response_

### Relevant log output

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by sadath-12**
* Remove `tetragon_map_drops_total`
* Remove `tetragon_map_in_use_gauge{map="eventcache"}`
* Replace `tetragon_map_in_use_gauge{map="processLru"}` with `tetragon_process_cache_size`
* Improve doc comments and help texts for some metrics

```release-note
Fix metrics related to monitoring BPF maps and userspace caches. Remove `tetragon_map_drops_total` (it was duplicating `tetragon_errors_total{type="process_cache_evicted"}`). Remove `tetragon_map_in_use_gauge{map="eventcache"}` (event cache is not a BPF map). Replace `tetragon_map_in_use_gauge{map="processLru"}` with `tetragon_process_cache_size` (process cache is not a BPF map).
```

Ref #1774

### CODE REVIEW

**Comment by sadath-12** [comment]
WIP !

**Comment by kkourt** [comment]
Moved to draft since it seems to be WIP.

**Comment by sadath-12** [comment]
> Moved to draft since it seems to be WIP.

yes thank you @kkourt

**Comment by sadath-12** [comment]
ready to review

**Comment by lambdanis** [comment]
@sadath-12 I see your PR has conflicts with the main branch, can you rebase?

**Comment by sadath-12** [comment]
Done @lambdanis

**Comment by lambdanis** [comment]
@sadath-12 It seems commits are still a bit off. You can check this in the checkpath action run ([here](https://github.com/cilium/tetragon/actions/runs/7479940809/job/20358229619?pr=1950).

There should be no merge commits, and ideally different tasks you implemented in this PR should be separate commits - they're quite independent of each other, so separate commits would make it easier to review the git history. I think it would be good to rewrite the git history here:
```
git reset HEAD~2
git add -p
# add only changes for a single task
git commit
# repeat git add, git commit for other tasks
```

**Comment by sadath-12** [comment]
@lambdanis  thank you . But I get 100's of files  in staged section when I run the reset command . Any efficient solution for this  please ?

**Comment by sadath-12** [comment]
Hi @lambdanis  in order to resolve the conflict . I pull this branch and do ```git merge --squash main``` and then ```git rebase -i HEAD~2``` when i enter the first interactive terminal to squash the 2nd commit and then 2nd terminal to write my final commit I see all the merged commits information which is very hard to remove them one by 1 . there are 100's of them . Any solution

**Comment by willfindlay** [comment]
Please fix the merge commit to have a proper description. Also `<90619459+sadath-12@users.noreply.github.com>` seems to be messing up our signed-off-by detection. Could you replace it with your email like you did for the other commit?

**Comment by willfindlay** [comment]
Needs formatting with `goimports`.

**Comment by sadath-12** [comment]
Hi I tried arranging the commits properly but it seems to be a mess . I tried various methods such as resetting to my original commit and then doing ```merge-squash``` . I also tried what @lambdanis  suggested , to reset and undo the merges I did that but when I push It does not sync with main branch and asks me to resolve conflicts

**Comment by lambdanis** [comment]
@sadath-12 Could you split the changes into logical commits? As there are five different fixes, it would make sense to have five separate commits for them in this case.

File: pkg/metrics/eventcachemetrics/eventcachemetrics.go [hunk 1]
```diff
@@ -39,6 +41,12 @@ var (
 		Help:        "The total of errors encountered while fetching process exec information from the cache.",
 		ConstLabels: nil,
 	}, []string{"error"})
+	ProcessCacheTotal = prometheus.NewCounterVec(prometheus.CounterOpts{
```
Comment (lambdanis) [inline pkg/metrics/eventcachemetrics/eventcachemetrics.go:44]: It would make sense to name the metric `event_cache_size` I think, and make it a gauge instead of counter. Using a counter here might be not incorrect if the cache size is only increasing until it's reset, but that's an implementation detail of the cache.

File: pkg/metrics/eventcachemetrics/eventcachemetrics.go [hunk 2]
```diff
@@ -39,6 +41,12 @@ var (
...
+		Namespace:   consts.MetricsNamespace,
+		Name:        "event_cache_process_total",
+		Help:        "The size of the process cache.",
+		ConstLabels: nil,
+	}, []string{"type"})
```
Comment (lambdanis) [inline pkg/metrics/eventcachemetrics/eventcachemetrics.go:49]: I don't understand what's the "type" here? I would expect this metric to have no labels.

**Comment by sadath-12** [comment]
Hi @lambdanis  . If there are 5 commits and let's say there was a review on the 3rd commit and pushed the new changes by squashing would it reflect on the 3rd commit ? . Requesting you to accept this in 1 commit itself as they are very small changes and I have given explanation on description

File: pkg/metrics/eventcachemetrics/eventcachemetrics.go (hunk 2 above)
Comment (sadath-12) [reply pkg/metrics/eventcachemetrics/eventcachemetrics.go:49]: did it because in ```FreeCache()``` pkg/process/process.go we had to reset and I think the counter would not provide the reset and If I remember I saw in some github issue for that usecase recommended to use it like this as it would provide reset option

**Comment by willfindlay** [comment]
> Requesting you to accept this in 1 commit itself as they are very small changes and I have given explanation on description

Could you reflect this description in the commit message body? It's empty at the moment.

**Comment by willfindlay** [comment]
@sadath-12 also it should be relatively easy to split this into separate commits. Here's how you can do it: 
0. Make a backup copy of your branch with `git branch -c backup`
1. `git reset --soft HEAD~1` to remove the existing commit but keep the changes.
2. `git add -p`
3. Select the changes you want to include in commit 1 following the instructions on the screen.
4. When you're done, write the commit message.
5. Repeat step 2-4 until there are no changes left.
6. When you're happy with it, force push.

**Comment by sadath-12** [comment]
Ya I can separate the commits like 1,2,3,4 . if after review I get changed related to 2nd commit how I make a new change and squash it under 2 ?

**Comment by lambdanis** [comment]
> Ya I can separate the commits like 1,2,3,4 . if after review I get changed related to 2nd commit how I make a new change and squash it under 2 ?

@sadath-12 Create a temporary commit, then use `git rebase -i HEAD~4` (interactive mode which allows you to rewrite the git history - reorder, squash, amend previous commits).

File: pkg/process/cache.go [hunk 3]
```diff
@@ -132,7 +131,7 @@ func NewCache(
 	lruCache, err := lru.NewWithEvict(
 		processCacheSize,
 		func(_ string, _ *ProcessInternal) {
-			mapmetrics.MapDropInc("processLru")
+			errormetrics.ErrorTotalInc(errormetrics.ProcessCacheEvicted)
```
[the remark this answers is not in the document]
Comment (sadath-12) [reply pkg/process/cache.go:134]: but in both cases evictions can happen right ?

[the remark this answers is not in the document]
Comment (lambdanis) [reply pkg/process/cache.go:134]: My understanding is that the eviction callback (defined here) is called on every eviction, so incrementing the metric in the `add` method too duplicates it. I would increment the metric only in the eviction callback. cc @willfindlay to confirm.

[the remark this answers is not in the document]
Comment (sadath-12) [reply pkg/process/cache.go:134]: ok by looking at the code from first point I feel whenever we initiate the cache with NewCache with some item and then maybe add other items using ```add``` so both cases eviction can happen . so that's where I thought both place we might consider to add this condition . let me know if I am wrong

[the remark this answers is not in the document]
Comment (willfindlay) [reply pkg/process/cache.go:134]: @lambdanis is correct. We should remove the other increment and only count it here.

**Comment by sadath-12** [comment]
> g too l

ya reworded that . Tried to explain in description without making much sense (but enough) in the title of commit msg

**Comment by sadath-12** [comment]
> @sadath-12 I see the CI is failing because of a commit message being too long: https://github.com/cilium/tetragon/actions/runs/8004683889/job/21862554217?pr=1950
> 
> Could you rephrase it to fit the characters limit?

ya reworded that . Tried to explain in description without making much sense (but enough) in the title of commit msg

**Comment by jrfastab** [comment]
@lambdanis @willfindlay could you take a look and merge if its ready to go? Thanks. CI failure there is something unrelated, I see this failure from time to time.

File: pkg/metrics/mapmetrics/mapmetrics.go [hunk 4]
```diff
@@ -10,33 +10,22 @@ import (
...
 	MapSize = metrics.NewBPFGauge(prometheus.NewDesc(
-		prometheus.BuildFQName(consts.MetricsNamespace, "", "map_in_use_gauge"),
+		prometheus.BuildFQName(consts.MetricsNamespace, "", "map_in_use"),
 		"The total number of in-use entries per map.",
 		[]string{"map", "total"}, nil,
 	))
 	MapErrors = metrics.NewBPFCounter(prometheus.NewDesc(
 		prometheus.BuildFQName(consts.MetricsNamespace, "", "map_errors_total"),
-		"The total number of entries dropped per LRU map.",
+		"The errors dropped per map.",
```
Comment (lambdanis) [inline pkg/metrics/mapmetrics/mapmetrics.go:20]:
```suggestion
		"The number of errors per map.",
```
This metric is just counting errors, nothing dropped.

**Review by lambdanis — CHANGES_REQUESTED** [review]
One request for a fix in a metrics description, otherwise looks good.

**Comment by lambdanis** [comment]
Thanks @sadath-12 for the updates. The code changes look good now. I rewrote the commit messages to describe exactly which metrics changed and split some commits to not mix up different changes. I dropped the `_gauge` suffix removal, as it's a breaking change that's hard to justify as a bug, so it shouldn't be mixed up in one PR with unrelated changes IMO. I also rewrote the PR description to include a release note.

**Review by lambdanis — APPROVED** [review]

**Comment by lambdanis** [comment]
The CI failure is a known issue: https://github.com/cilium/tetragon/issues/2010. I'm merging this PR.
